File: farmmarket_mobile/lib/models/crop.dart
class Crop {
  final int id;
  final String name;
  final String description;
  final double pricePerKg;
  final double availableQuantity;
  final String unit;
  final String category;
  final String region;
  final String? image;
  final bool isAvailable;
  final String createdAt;
  final String updatedAt;
  final Farmer farmer;

  Crop({
    required this.id,
    required this.name,
    required this.description,
    required this.pricePerKg,
    required this.availableQuantity,
    required this.unit,
    required this.category,
    required this.region,
    this.image,
    required this.isAvailable,
    required this.createdAt,
    required this.updatedAt,
    required this.farmer,
  });

  factory Crop.fromJson(Map<String, dynamic> json) {
    return Crop(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      pricePerKg: double.parse(json['price_per_kg'].toString()),
      availableQuantity: double.parse(json['available_quantity'].toString()),
      unit: json['unit'] ?? '',
      category: json['category'] ?? '',
      region: json['region'] ?? '',
      image: json['image'],
      isAvailable: json['is_available'] ?? false,
      createdAt: json['created_at'] ?? '',
      updatedAt: json['updated_at'] ?? '',
      farmer: Farmer.fromJson(json['farmer'] ?? {}),
    );
  }
}

class Category {
  final int id;
  final String name;
  final String description;
  final int cropsCount;

  Category({
    required this.id,
    required this.name,
    required this.description,
    required this.cropsCount,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      name: json['name'],
      description: json['description'] ?? '',
      cropsCount: json['crops_count'] ?? 0,
    );
  }
}

class Farmer {
  final int id;
  final String name;
  final String email;
  final String phone;
  final bool isVerified;

  Farmer({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
    required this.isVerified,
  });

  factory Farmer.fromJson(Map<String, dynamic> json) {
    return Farmer(
      id: json['id'],
      name: json['name'],
      email: json['email'],
      phone: json['phone'],
      isVerified: json['is_verified'] ?? false,
    );
  }
}

class CropImage {
  final int id;
  final String imagePath;

  CropImage({
    required this.id,
    required this.imagePath,
  });

  factory CropImage.fromJson(Map<String, dynamic> json) {
    return CropImage(
      id: json['id'],
      imagePath: json['image_path'],
    );
  }

  String get fullImageUrl => 'http://127.0.0.1:8000/storage/$imagePath';
}

File: app/Models/Crop.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Crop extends Model
{
    use HasFactory;

    protected $fillable = [
        'farmer_id',
        'name',
        'description',
        'price_per_kg',
        'available_quantity',
        'unit',
        'category',
        'region',
        'image',
        'is_available',
    ];

    protected $casts = [
        'price_per_kg' => 'decimal:2',
        'available_quantity' => 'decimal:2',
        'is_available' => 'boolean',
    ];

    /**
     * Get the farmer that owns the crop
     */
    public function farmer()
    {
        return $this->belongsTo(User::class, 'farmer_id');
    }

    /**
     * Get the order items for this crop
     */
    public function orderItems()
    {
        return $this->hasMany(OrderItem::class);
    }

    /**
     * Scope a query to only include available crops
     */
    public function scopeAvailable($query)
    {
        return $query->where('is_available', true);
    }

    /**
     * Scope a query to filter by category
     */
    public function scopeByCategory($query, $category)
    {
        return $query->where('category', $category);
    }

    /**
     * Scope a query to filter by region
     */
    public function scopeByRegion($query, $region)
    {
        return $query->where('region', $region);
    }

    /**
     * Scope a query to filter by price range
     */
    public function scopeByPriceRange($query, $minPrice, $maxPrice)
    {
        return $query->whereBetween('price_per_kg', [$minPrice, $maxPrice]);
    }

    /**
     * Get the formatted price
     */
    public function getFormattedPriceAttribute()
    {
        return 'TZS' . number_format($this->price_per_kg, 2);
    }

    /**
     * Get the image URL
     */
    public function getImageUrlAttribute()
    {
        if ($this->image) {
            return asset('storage/crops/' . $this->image);
        }
        return asset('images/default-crop.jpg');
    }

    /**
     * Check if crop has active orders
     */
    public function hasActiveOrders()
    {
        return $this->orderItems()
            ->whereHas('order', function($query) {
                $query->whereIn('status', ['pending', 'confirmed']);
            })
            ->exists();
    }
}
